Use clang-style USRs for swift decls that are exposed to Objective C

nathawes opened the pull request:
<!-- What's in this pull request? -->
We have an issue where the Swift symbols that are exported to ObjC via the generated ObjC header that the Swift compiler emits, have different USRs when indexed from swift & clang.
The swift side uses Swift-specific USRs while the clang side uses it’s own ObjC USRs, making these symbols and their references seem as if they are different symbols (e.g. for jump-to-definition, callers, etc).

This patch changes the USR generation for @objc swift symbols to use clang-style USRs, so there is a single USR cross-language.

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
This is a step towards resolving rdar://problem/16271632, but it doesn't yet 'namespace' the generated clang-style USRs with the swift module name (waiting on some clang changes).  This is needed to avoid USR conflicts with symbols from different Swift frameworks.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

Commit:
Use clang-style USRs for swift decls that are exposed to Objective C

nathawes commented:
@swift-ci please smoke test

benlangmuir commented on lib/AST/Decl.cpp:
```diff
@@ -5042,3 +5042,20 @@ getObjCNameForSwiftDecl(const ValueDecl *VD, DeclName PreferredName){
+    return true;
+  } else if (checkParent) {
+    if (auto ctor = dyn_cast<ConstructorDecl>(VD)) {
+      // Check if we're overriding an initializer that is visible to obj-c
```
Why is this the only case that checks the parent?

nathawes replied:
This is the logic PrintAsObjC was using to decide whether to include a decl in the generated header – I just moved the function here to make sure USR generation and the header printer use the same logic. I assume 'checkParent' was added specifically for the case of still writing out < minRequiredAccess initializers that override a >= minRequiredAccess initializer.

slavapestov replied:
If the overridden decl responds with true to isObjC(), the decl itself will, too. There's no need to check it explicitly. Also please call it the override, not the parent.

benlangmuir commented on lib/AST/USRGeneration.cpp:
```diff
@@ -45,6 +46,77 @@ bool ide::printDeclTypeUSR(const ValueDecl *D, raw_ostream &OS) {
+
+  auto Parent = D->getDeclContext()->getInnermostDeclarationDeclContext();
+  if (Parent && (!ShouldUseObjCName(Parent) || // parent should be visible too
+                    !D->getDeclContext()->isTypeContext() || // no local decls
```
Indentation is off.

benlangmuir commented on lib/AST/USRGeneration.cpp:
```diff
@@ -45,6 +46,77 @@ bool ide::printDeclTypeUSR(const ValueDecl *D, raw_ostream &OS) {
+
+  if (const ValueDecl *VD = dyn_cast<ValueDecl>(D)) {
+    if (isa<EnumElementDecl>(VD))
+      return Parent;
```
Does an EnumElementDecl ever not have a parent?

nathawes replied:
Wasn't sure about invalid code, but it doesn't look like it – I'll change it to true ;-)

benlangmuir commented on test/IDE/print_usrs.swift:
```diff
@@ -149,13 +149,91 @@ class Observers {
+  }
+
+  // CHECK: [[@LINE+1]]:14 c:objc(cs)ObjCClass1(cpy)typeComputed{{$}}
+  static var typeComputed: Int {
```
Does this actually show up as a class property in ObjC?  I would have expected you'd need to say "class" instead of "static" (they're not the same).

nathawes replied:
It does at the moment, yeah:
SWIFT_CLASS_PROPERTY(@property (nonatomic, class) NSInteger typeComputed;)
+ (NSInteger)typeComputed SWIFT_WARN_UNUSED_RESULT;
+ (void)setTypeComputed:(NSInteger)newValue;

nkcsgexi commented on include/swift/AST/SwiftNameTranslation.h:
```diff
@@ -44,6 +44,11 @@ namespace objc_translation {
+  /// Returns true if the given value decl D is visible to ObjC of its
+  /// own accord (i.e. without considering its context)
+  bool isVisibleToObjC(const ValueDecl *VD, Accessibility minRequiredAccess,
+                     bool checkParent = true);
```
nitpick: `bool` should align with `const` in the previous line.

nkcsgexi commented on lib/AST/Decl.cpp:
```diff
@@ -5042,3 +5042,20 @@ getObjCNameForSwiftDecl(const ValueDecl *VD, DeclName PreferredName){
+  } else if (checkParent) {
+    if (auto ctor = dyn_cast<ConstructorDecl>(VD)) {
+      // Check if we're overriding an initializer that is visible to obj-c
+      if (auto parent = ctor->getOverriddenDecl())
```
do we care about `getSatisfiedProtocolRequirements()` here?

nkcsgexi commented on lib/AST/USRGeneration.cpp:
```diff
@@ -45,6 +46,77 @@ bool ide::printDeclTypeUSR(const ValueDecl *D, raw_ostream &OS) {
+    OS << "@" << ObjCName;
+  } else {
+    llvm_unreachable("Unexpected value decl");
+    return true;
```
nitpick: redundant `return` here

nkcsgexi commented on lib/AST/USRGeneration.cpp:
```diff
@@ -45,6 +46,77 @@ bool ide::printDeclTypeUSR(const ValueDecl *D, raw_ostream &OS) {
+  OS << clang::index::getUSRSpacePrefix();
+
+  if (D->getDeclContext()->isTypeContext()) {
+    auto *Parent = D->getDeclContext()->getAsNominalTypeOrNominalTypeExtensionContext();
```
these above two lines can be combined as:
`if (auto *Parent = D->getDeclContext()->getAsNominalTypeOrNominalTypeExtensionContext())`

nkcsgexi commented on lib/AST/USRGeneration.cpp:
```diff
@@ -45,6 +46,77 @@ bool ide::printDeclTypeUSR(const ValueDecl *D, raw_ostream &OS) {
+
+  if (!Ident.empty())
+    return printObjCNameFragment(D, Ident.str(), OS);
+  if (Selector) {
```
`assert(Selector)` here. one and only one of `Ident` and `Selector` is valid.

nathawes replied:
Are there no cases (e.g. invalid code) where they'll both come back invalid?

nkcsgexi replied:
For invalid code, we may get nameless value decls. Can we have an early exist in this function like `if (VD->getName().empty) return false` like the one in `ide::printDeclUSR` so that we know what's going on without ignoring the cases?

nathawes replied:
Sounds good!

nkcsgexi replied:
I realize we do have similar forgiving check in `SwiftSourceDocInfo.cpp`. The reason there is that we may have preferred names that are non-translatable to ObjC names, for instance, when the argument counts do not match. However, these cases are unlikely to happen here because we don't have preferred names.

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -5042,3 +5042,20 @@ getObjCNameForSwiftDecl(const ValueDecl *VD, DeclName PreferredName){
 }
+
+bool swift::objc_translation::
+isVisibleToObjC(const ValueDecl *VD, Accessibility minRequiredAccess,
```
Is there a better place for this than Decl.cpp?

nkcsgexi replied:
We want the utilities to be inside libAST. Which file do you suggest?

slavapestov replied:
I would suggest either moving it to some other file, in lib/IDE/, or making it a method on ValueDecl instead of a top-level function. Just feels weird to see 'swift::objc_translation' in this file.

slavapestov replied:
Also how about 'isVisibleInBridgingHeader' instead of 'isVisibleToObjC'?

benlangmuir replied:
Bridging header is the other direction (ObjC code visible in Swift).

akyrtzi replied:
It's generally preferable for headers to have an associated implementation file, so `SwiftNameTranslation.h` can have `SwiftNameTranslation.cpp` for its code, it doesn't have to go to `Decl.cpp`

slavapestov replied:
I like @akyrtzi's suggest the best.

Commit:
Move SwiftNameTranslation implementation out of Decl.cpp and into its own file.
Also fix code formatting issues and simplify the code in USRGeneration.cpp based on review comments in PR #7670.

nathawes commented:
@swift-ci please smoke test

Commit:
Fix bad indentation in SwiftNameTranslation.h

nathawes commented:
@swift-ci please smoke test

nkcsgexi commented on lib/AST/SwiftNameTranslation.cpp:
```diff
@@ -0,0 +1,117 @@
+                bool checkParent) {
+  if (!(VD->isObjC() || VD->getAttrs().hasAttribute<CDeclAttr>()))
+    return false;
+  if (VD->getFormalAccess() >= minRequiredAccess) {
```
can we check `VD->hasFormalAccess()` before getting it? invalid code usually has this trap.

slavapestov replied:
You mean hasAccessibility()?

nkcsgexi replied:
Right

Commit:
SwiftNameTranslation.cpp: Guard call to VD->getFormalAccess() with a hasAccessibility() check.

nathawes commented:
@swift-ci please smoke test
